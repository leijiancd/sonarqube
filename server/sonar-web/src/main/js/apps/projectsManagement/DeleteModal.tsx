/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
import * as React from 'react';
import Modal from 'react-modal';
import { deleteComponents } from '../../api/components';
import { translate } from '../../helpers/l10n';

export interface Props {
  onClose: () => void;
  onConfirm: () => void;
  organization: string;
  qualifier: string;
  selection: string[];
}

interface State {
  loading: boolean;
}

export default class DeleteModal extends React.PureComponent<Props, State> {
  mounted: boolean;
  state: State = { loading: false };

  componentDidMount() {
    this.mounted = true;
  }

  componentWillUnmount() {
    this.mounted = false;
  }

  handleCancelClick = (event: React.SyntheticEvent<HTMLAnchorElement>) => {
    event.preventDefault();
    this.props.onClose();
  };

  handleConfirmClick = () => {
    this.setState({ loading: true });
    deleteComponents(this.props.selection, this.props.organization).then(
      () => {
        if (this.mounted) {
          this.props.onConfirm();
        }
      },
      () => {
        if (this.mounted) {
          this.setState({ loading: false });
        }
      }
    );
  };

  render() {
    const header = translate('qualifiers.delete', this.props.qualifier);

    return (
      <Modal
        isOpen={true}
        contentLabel={header}
        className="modal"
        overlayClassName="modal-overlay"
        onRequestClose={this.props.onClose}>
        <header className="modal-head">
          <h2>
            {header}
          </h2>
        </header>

        <div className="modal-body">
          {translate('qualifiers.delete_confirm', this.props.qualifier)}
        </div>

        <footer className="modal-foot">
          {this.state.loading && <i className="spinner spacer-right" />}
          <button
            className="button-red"
            disabled={this.state.loading}
            onClick={this.handleConfirmClick}>
            {translate('delete')}
          </button>
          <a className="js-modal-close" href="#" onClick={this.handleCancelClick}>
            {translate('cancel')}
          </a>
        </footer>
      </Modal>
    );
  }
}
