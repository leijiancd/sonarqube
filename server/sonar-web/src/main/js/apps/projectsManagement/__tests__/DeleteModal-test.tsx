/*
 * SonarQube
 * Copyright (C) 2009-2016 SonarSource SA
 * mailto:contact AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Lesser General Public
 * License as published by the Free Software Foundation; either
 * version 3 of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Lesser General Public License for more details.
 *
 * You should have received a copy of the GNU Lesser General Public License
 * along with this program; if not, write to the Free Software Foundation,
 * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 */
jest.mock('../../../api/components', () => ({
  deleteComponents: jest.fn(() => Promise.resolve())
}));

import * as React from 'react';
import { shallow } from 'enzyme';
import DeleteModal, { Props } from '../DeleteModal';
import { click } from '../../../helpers/testUtils';

const deleteComponents = require('../../../api/components').deleteComponents as jest.Mock<any>;

it('deletes projects', async () => {
  const onConfirm = jest.fn();
  const wrapper = shallowRender({ onConfirm });
  (wrapper.instance() as DeleteModal).mounted = true;
  expect(wrapper).toMatchSnapshot();

  click(wrapper.find('button'));
  expect(wrapper).toMatchSnapshot();
  expect(deleteComponents).toBeCalledWith(['foo', 'bar'], 'org');

  await new Promise(setImmediate);
  expect(onConfirm).toBeCalled();
});

it('closes', () => {
  const onClose = jest.fn();
  const wrapper = shallowRender({ onClose });
  click(wrapper.find('.js-modal-close'));
  expect(onClose).toBeCalled();
});

function shallowRender(props?: { [P in keyof Props]?: Props[P] }) {
  return shallow(
    <DeleteModal
      onClose={jest.fn()}
      onConfirm={jest.fn()}
      organization="org"
      qualifier="TRK"
      selection={['foo', 'bar']}
      {...props}
    />
  );
}
