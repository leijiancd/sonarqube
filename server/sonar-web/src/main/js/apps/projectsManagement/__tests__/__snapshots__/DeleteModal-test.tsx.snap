// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deletes projects 1`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="qualifiers.delete.TRK"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      qualifiers.delete.TRK
    </h2>
  </header>
  <div
    className="modal-body"
  >
    qualifiers.delete_confirm.TRK
  </div>
  <footer
    className="modal-foot"
  >
    <button
      className="button-red"
      disabled={false}
      onClick={[Function]}
    >
      delete
    </button>
    <a
      className="js-modal-close"
      href="#"
      onClick={[Function]}
    >
      cancel
    </a>
  </footer>
</Modal>
`;

exports[`deletes projects 2`] = `
<Modal
  ariaHideApp={true}
  bodyOpenClassName="ReactModal__Body--open"
  className="modal"
  closeTimeoutMS={0}
  contentLabel="qualifiers.delete.TRK"
  isOpen={true}
  onRequestClose={[Function]}
  overlayClassName="modal-overlay"
  parentSelector={[Function]}
  portalClassName="ReactModalPortal"
  shouldCloseOnOverlayClick={true}
>
  <header
    className="modal-head"
  >
    <h2>
      qualifiers.delete.TRK
    </h2>
  </header>
  <div
    className="modal-body"
  >
    qualifiers.delete_confirm.TRK
  </div>
  <footer
    className="modal-foot"
  >
    <i
      className="spinner spacer-right"
    />
    <button
      className="button-red"
      disabled={true}
      onClick={[Function]}
    >
      delete
    </button>
    <a
      className="js-modal-close"
      href="#"
      onClick={[Function]}
    >
      cancel
    </a>
  </footer>
</Modal>
`;
